x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -1165,8 +1213,134 @@ protected LogicalPlan rule(Aggregate aggregate) {
             return aggregate;
         }
 
-        static String temporaryName(NamedExpression agg, AggregateFunction af) {
-            return SubstituteSurrogates.temporaryName(agg, af);
+        static String temporaryName(Expression expression, AggregateFunction af) {
+            return SubstituteSurrogates.temporaryName(expression, af);
+        }
+    }
+
+    /**
+     * Replace nested expressions over aggregates with synthetic eval post the aggregation
+     * stats a = sum(a) + min(b) by x
+     * becomes
+     * stats a1 = sum(a), a2 = min(b) by x | eval a = a1 + a2 | keep a, x
+     *
+     * Since the logic is very similar, this rule also handles duplicate aggregate functions to avoid duplicate compute
+     * stats a = min(x), b = min(x), c = count(*), d = count() by g
+     * becomes
+     * stats a = min(x), c = count(*) by g | eval b = a, d = c | keep a, b, c, d, g
+     */
+    static class ReplaceStatsAggExpressionWithEval extends OptimizerRules.OptimizerRule<Aggregate> {
+        ReplaceStatsAggExpressionWithEval() {
+            super(TransformDirection.UP);
+        }
+
+        @Override
+        protected LogicalPlan rule(Aggregate aggregate) {
+            // build alias map
+            AttributeMap<Expression> aliases = new AttributeMap<>();
+            aggregate.forEachExpressionUp(Alias.class, a -> aliases.put(a.toAttribute(), a.child()));
+
+            // break down each aggregate into AggregateFunction
+            // preserve the projection at the end
+            Holder<Boolean> aggsChanged = new Holder<>(false);
+            List<? extends NamedExpression> aggs = aggregate.aggregates();
+
+            // root/naked aggs
+            Map<AggregateFunction, Alias> rootAggs = Maps.newLinkedHashMapWithExpectedSize(aggs.size());
+            // evals (original expression relying on multiple aggs)
+            List<Alias> newEvals = new ArrayList<>();
+            List<NamedExpression> newProjections = new ArrayList<>();
+            // track the aggregate aggs (including grouping which is not an AggregateFunction)
+            List<NamedExpression> newAggs = new ArrayList<>();
+
+            Holder<Boolean> changed = new Holder<>(false);
+
+            for (NamedExpression agg : aggs) {
+                if (agg instanceof Alias as) {
+                    // if the child a nested expression
+                    Expression child = as.child();
+
+                    // common case - handle duplicates
+                    if (child instanceof AggregateFunction af) {
+                        AggregateFunction canonical = (AggregateFunction) af.canonical();
```

Comment:
Interesting. Wondering if it's worth double checking this in a test of the form `stats avg(a+b), avg(b+a)` (using the canonical form of commutative operators).

Reply:
Added a test for it just in case.